### ISSUE TITLE
Support Qt5

### ISSUE THREAD

**Issue by MicBosi** [MicBosi/VisualizationLibrary#81]
Looks like most users use VL via Qt so as part of the current refocusing effort VL will update support to the latest GUI back ends but pay special attention to Qt5 as its primary GUI.

**Comment by Scott-Merritt** [comment]
I'm very glad to hear that this project will continue to be maintained. I've been using VL in one of my projects for a few years, and have upgraded to Qt5 during that time. I modified VL's CMake build system in order to allow VL to compile against Qt5 and am happy to share those modifications with you if it would save you time.  

Note that this will break Qt4 (since I had no interest in maintining Qt4 functionality)

The diff is below.

```
diff --git src/CMakeLists.txt src/CMakeLists.txt
index 9d3fb84..584eadc 100644
--- src/CMakeLists.txt
+++ src/CMakeLists.txt
@@ -53,6 +53,21 @@ if(VL_GUI_QT4_SUPPORT)
 	add_subdirectory("vlQt4")
 endif()
 
+# Qt5
+option(VL_GUI_QT5_SUPPORT "Build Qt5 support" ON)
+if(VL_GUI_QT5_SUPPORT)
+	find_package(Qt5Core REQUIRED)
+	find_package(Qt5Gui REQUIRED)
+	find_package(Qt5OpenGL REQUIRED)
+	find_package(Qt5Widgets REQUIRED)
+	include_directories("${Qt5Core_INCLUDE_DIRS}")
+	include_directories("${Qt5Gui_INCLUDE_DIRS}")
+	include_directories("${Qt5OpenGL_INCLUDE_DIRS}")
+	include_directories("${Qt5Widgets_INCLUDE_DIRS}")
+	add_subdirectory("vlQt4")
+endif()
+
+
 if(VL_PLATFORM_WINDOWS)
 	option(VL_GUI_WIN32_SUPPORT "Build Win32 support" OFF)
 endif()
diff --git src/vlQt4/CMakeLists.txt src/vlQt4/CMakeLists.txt
index 4945510..ade0bbd 100644
--- src/vlQt4/CMakeLists.txt
+++ src/vlQt4/CMakeLists.txt
@@ -21,7 +21,11 @@ project(VLQt4)
 file(GLOB VLQt4_SRC "*.cpp")
 file(GLOB VLQt4_INC "*.hpp")
 
+if(VL_GUI_QT4_SUPPORT)
 QT4_WRAP_CPP(MOC_SOURCES Qt4Widget.hpp)
+elseif(VL_GUI_QT5_SUPPORT)
+QT5_WRAP_CPP(MOC_SOURCES Qt4Widget.hpp)
+endif()
 
 add_library(VLQt4 ${VL_SHARED_OR_STATIC} ${VLQt4_SRC} ${VLQt4_INC} ${MOC_SOURCES})
 VL_DEFAULT_TARGET_PROPERTIES(VLQt4)
diff --git src/vlQt4/Qt4Widget.cpp src/vlQt4/Qt4Widget.cpp
index bb42f86..4197965 100644
--- src/vlQt4/Qt4Widget.cpp
+++ src/vlQt4/Qt4Widget.cpp
@@ -32,6 +32,7 @@
 #include <vlQt4/Qt4Widget.hpp>
 #include <vlCore/Log.hpp>
 
+
 using namespace vl;
 using namespace vlQt4;
 
diff --git src/vlQt4/Qt4Widget.hpp src/vlQt4/Qt4Widget.hpp
index 1b2d062..f363d5c 100644
--- src/vlQt4/Qt4Widget.hpp
+++ src/vlQt4/Qt4Widget.hpp
@@ -35,14 +35,15 @@
 #include <vlQt4/link_config.hpp>
 #include <vlCore/VisualizationLibrary.hpp>
 #include <vlGraphics/OpenGLContext.hpp>
-#include <QtGui/QApplication>
-#include <QtGui/QMouseEvent>
-#include <QtGui/QWidget>
-#include <QtCore/QUrl>
-#include <QtCore/QTimer>
-#include <QtCore/QObject>
-#include <QtOpenGL/QGLWidget>
-#include <QtOpenGL/QGLFormat>
+#include <QtWidgets>
+#include <QApplication>
+#include <QMouseEvent>
+#include <QWidget>
+#include <QUrl>
+#include <QTimer>
+#include <QObject>
+#include <QGLWidget>
+#include <QGLFormat>
 
 namespace vlQt4
 {
```

**Comment by MicBosi** [comment]
Awesome thank you. If possible I will try to keep compatibility with both Qt4 and Qt5.

**Comment by reso** [comment]
Very glad to see activity on visualization library, I know Qt5 (QWidgets) support would be very welcome and useful feature.  I am watching the repository and can try to help with testing and possibly rewriting/porting some of the previous examples and tutorials to Qt5 when appropriate.

**Comment by jdgibbs** [comment]
I also want to echo my thanks for the continued support and development of VL.  I'm also using VL in my projects within Qt.  Aside from Qt5 support, it would be great if VL/OpenGL would play nicer with other Qt drawing methods.  In particular, I've had difficulty getting QPainter to work on top of VL rendered scenes.  It would often look fine, but then seemingly random characters within a string would occasionally not render.  I've found it sometime easier to overlay from Qt and not VL, if that were a reliable option.  

An even bigger request would be to see QML and VL co-operate.  It'd be really great to put VL directly into QML layouts, and/or layer QML controls over a VL scene.

I realize much of the limitations may be with QT rather than VL but while this project seems to be in the "wish list" phase I felt compelled to add to it.

I'd be happy to help test and contribute as the project moves forward.  Thank you again for your fine work!

**Comment by MicBosi** [comment]
Thanks, I'll do my best to move things forward:)

I had a quick look at QML/QtQuick/OpenGL, we're having a little discussion here in case you missed it:
https://groups.google.com/forum/#!topic/visualization-library/izUCZoCnAi8

**Comment by MicBosi** [comment]
Hi everyone, I've just released v2.0-beta which includes Qt5 as well as improved Qt4 support among other things, feel free to test and send Pull Requests against the current master if you have goodies to share :)

cc @jdgibbs @reso @Scott-Merritt
